src/server/services/wfs3/qgswfs3handlers.cpp

```diff
@@ -0,0 +1,1178 @@
+    json selfLink;
+    for ( const auto &l : data["links"] )
+    {
+      if ( l["rel"] == "self" )
+      {
+        selfLink = l;
+        break;
+      }
+    }
+    // This should never happen!
+    Q_ASSERT( !selfLink.is_null() );
+
+    // Add prev - next links
+    if ( offset != 0 )
+    {
+      auto prevLink { selfLink };
+      prevLink["href"] = QStringLiteral( "%1&offset=%2&limit=%3" ).arg( cleanedUrl ).arg( std::max<long>( 0, limit - offset ) ).arg( limit ).toStdString();
+      prevLink["rel"] = "prev";
+      prevLink["name"] = "Previous page";
+      data["links"].push_back( prevLink );
+    }
+    if ( limit + offset < matchedFeaturesCount )
+    {
+      auto nextLink { selfLink };
+      nextLink["href"] = QStringLiteral( "%1&offset=%2&limit=%3" ).arg( cleanedUrl ).arg( std::min<long>( matchedFeaturesCount, limit + offset ) ).arg( limit ).toStdString();
+      nextLink["rel"] = "next";
+      nextLink["name"] = "Next page";
+      data["links"].push_back( nextLink );
+    }
+
+    json navigation = json::array();
+    navigation.push_back( {{ "title",  "Landing page" }, { "href", parentLink( url, 3 ) }} ) ;
+    navigation.push_back( {{ "title",  "Collections" }, { "href", parentLink( url, 2 ) }} ) ;
+    navigation.push_back( {{ "title",   title }, { "href", parentLink( url, 1 )  }} ) ;
+    json htmlMetadata
+    {
+      { "pageTitle", "Features in layer " + title },
+      { "layerTitle", title },
+      {
+        "geojsonUrl", href( context, "/",
+                            QgsServerOgcApi::contentTypeToExtension( QgsServerOgcApi::ContentType::GEOJSON ) )
+      },
+      { "navigation", navigation }
+    };
+    write( data, context, htmlMetadata );
+  }
+  else
+  {
+    throw QgsServerApiNotImplementedException( QStringLiteral( "Only GET method is implemented." ) );
+  }
+
+}
+
+Wfs3CollectionsFeatureHandler::Wfs3CollectionsFeatureHandler()
+{
+}
+
+void Wfs3CollectionsFeatureHandler::handleRequest( const QgsServerApiContext &context ) const
+{
+  if ( ! context.project() )
+  {
+    throw QgsServerApiImproperlyConfiguredException( QStringLiteral( "Project is invalid or undefined" ) );
+  }
+  // Check collectionId
+  const auto match { path().match( context.request()->url().path( ) ) };
+  if ( ! match.hasMatch() )
+  {
+    throw QgsServerApiNotFoundError( QStringLiteral( "Collection was not found" ) );
+  }
+  const auto collectionId { match.captured( QStringLiteral( "collectionId" ) ) };
+  // May throw if not found
+  const auto mapLayer { layerFromCollection( context, collectionId ) };
+  Q_ASSERT( mapLayer );
+  const auto title { mapLayer->title().isEmpty() ? mapLayer->name().toStdString() : mapLayer->title().toStdString() };
+
+  if ( context.request()->method() == QgsServerRequest::Method::GetMethod )
+  {
+    const auto featureId { match.captured( QStringLiteral( "featureId" ) ) };
+    QgsJsonExporter exporter { mapLayer };
+    auto feature { mapLayer->getFeature( featureId.toLongLong() ) };
```

Comment:
```suggestion
    const auto feature { mapLayer->getFeature( featureId.toLongLong() ) };
```

Reply:
To be honest, if I would change only a single thing here it would rather be to switch auto for QgsFeature than adding the const ;)